File: internal/models/clipping.go
package models

import (
	"encoding/json"
	"time"
)

// ClippingItem represents a single clipping from Kindle
type ClippingItem struct {
	Title     string    `json:"title"`
	Content   string    `json:"content"`
	PageAt    string    `json:"pageAt"`
	CreatedAt time.Time `json:"createdAt"`
}

// MarshalJSON implements custom JSON marshaling to maintain RFC3339 format
func (c ClippingItem) MarshalJSON() ([]byte, error) {
	type Alias ClippingItem
	return json.Marshal(&struct {
		*Alias
		CreatedAt string `json:"createdAt"`
	}{
		Alias:     (*Alias)(&c),
		CreatedAt: c.CreatedAt.UTC().Format(time.RFC3339),
	})
}

// UnmarshalJSON implements custom JSON unmarshaling to parse RFC3339 format
func (c *ClippingItem) UnmarshalJSON(data []byte) error {
	type Alias ClippingItem
	aux := &struct {
		*Alias
		CreatedAt string `json:"createdAt"`
	}{
		Alias: (*Alias)(c),
	}

	if err := json.Unmarshal(data, aux); err != nil {
		return err
	}

	var err error
	c.CreatedAt, err = time.Parse(time.RFC3339, aux.CreatedAt)
	return err
}

// ClippingInput represents the input format for GraphQL mutations
type ClippingInput struct {
	Title     string `json:"title"`
	Content   string `json:"content"`
	BookID    string `json:"bookID"`
	PageAt    string `json:"pageAt"`
	CreatedAt string `json:"createdAt"`
	Source    string `json:"source"`
}

// ToClippingInput converts ClippingItem to ClippingInput for API calls
func (c ClippingItem) ToClippingInput() ClippingInput {
	return ClippingInput{
		Title:     c.Title,
		Content:   c.Content,
		BookID:    "0", // Default book ID
		PageAt:    c.PageAt,
		CreatedAt: c.CreatedAt.UTC().Format(time.RFC3339),
		Source:    "kindle",
	}
}

File: src/models/clipping.ts
export interface ClippingItem {
  title: string;
  content: string;
  pageAt: string;
  createdAt: Date;
}

export interface ClippingInput {
  title: string;
  content: string;
  bookID: string;
  pageAt: string;
  createdAt: string;
  source: string;
}

export interface SerializedClipping {
  title: string;
  content: string;
  pageAt: string;
  createdAt: string;
}

export function toRFC3339(date: Date): string {
  return date.toISOString().replace(/\.\d{3}Z$/, "Z");
}

export function serializeClipping(c: ClippingItem): SerializedClipping {
  return {
    title: c.title,
    content: c.content,
    pageAt: c.pageAt,
    createdAt: toRFC3339(c.createdAt),
  };
}

export function toClippingInput(c: ClippingItem): ClippingInput {
  return {
    title: c.title,
    content: c.content,
    bookID: "0",
    pageAt: c.pageAt,
    createdAt: toRFC3339(c.createdAt),
    source: "kindle",
  };
}
